### PR TITLE
[pdex] Fix calculate virtual amount with out range uint64 - add liqudity

### PR DESCRIPTION
**Pull request by tin-incognito**
Fix calculate virtual amount with out range uint64 - add liqudity

### CODE REVIEW

File: blockchain/pdex/pool_pair_state.go [hunk 1]
```diff
@@ -453,7 +453,7 @@ func (p *PoolPairState) updateSingleTokenAmount(
 			tempVirtualAmount,
 			big.NewInt(0).SetUint64(p.state.ShareAmount()),
 		)
-		if tempVirtualAmount.Uint64() > newRealAmount {
+		if tempVirtualAmount.Cmp(big.NewInt(0).SetUint64(newRealAmount)) > 0 {
 			newVirtualAmount = tempVirtualAmount
 		} else {
 			newVirtualAmount.SetUint64(newRealAmount)
```
Comment (cuong-incognito) [inline blockchain/pdex/pool_pair_state.go:459]: Why does the virtual amount always have to be bigger than the real amount?

**Review by dat-incognito-org — APPROVED** [review]

File: blockchain/pdex/pool_pair_state.go (hunk 1 above)
Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go:459]: the virtual amount can not be smaller to newRealAmount

Comment (cuong-incognito) [reply blockchain/pdex/pool_pair_state.go:459]: where does the condition come from? Is it mentioned in DMM smart contract code or white paper?

Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go:459]: Based on theory amplifier will always be >= 1.0 and x' = a*x so x' should be always greater of equal to x
Based on practical: https://github.com/dynamic-amm/smart-contracts/blob/04b47545c58fa0ab99d4bb9eae729b3be31efb0c/contracts/DMMPool.sol#L106-L107

Comment (cuong-incognito) [reply blockchain/pdex/pool_pair_state.go:459]: If it always is, so what's the meaning of `newVirtualAmount.SetUint64(newRealAmount)` when checking `tempVirtualAmount < newRealAmount` if it never happens.

Comment (tin-incognito) [reply blockchain/pdex/pool_pair_state.go:459]: ok maybe you understand wrong here. This is full code block

```
newVirtualAmount := big.NewInt(0)
	if p.state.Amplifier() != metadataPdexv3.BaseAmplifier {
		tempVirtualAmount := big.NewInt(0).Mul(
			virtualAmount,
			big.NewInt(0).SetUint64(tempShareAmount),
		)
		tempVirtualAmount = tempVirtualAmount.Div(
			tempVirtualAmount,
			big.NewInt(0).SetUint64(p.state.ShareAmount()),
		)
		if tempVirtualAmount.Cmp(big.NewInt(0).SetUint64(newRealAmount)) > 0 {
			newVirtualAmount = tempVirtualAmount
		} else {
			newVirtualAmount.SetUint64(newRealAmount)
		}
	} else {
		newVirtualAmount, err = executeOperationBigInt(virtualAmount, big.NewInt(0).SetUint64(amount), operator)
	}
	switch tokenID.String() {
	case p.state.Token0ID().String():
		p.state.SetToken0RealAmount(newRealAmount)
		p.state.SetToken0VirtualAmount(newVirtualAmount)
	case p.state.Token1ID().String():
		p.state.SetToken1RealAmount(newRealAmount)
		p.state.SetToken1VirtualAmount(newVirtualAmount)
	}
``` 

so the newVirtualAmount is used for beneath

**Review by cuong-incognito — APPROVED** [review]

**Pull request merged**